Add API for component groups

4e6 opened the pull request:
### Pull Request Description

<!--
- Please describe the nature of your PR here, as well as the motivation for it.
- If it fixes an open issue, please mention that issue number here.
-->

PR ads two new commands to the Language Server protocol
- `library/getPackage` to get the package definition of the provided library
- `editions/listDefinedComponents` to get the list of available components in the provided edition

### Important Notes

The `library/getPackage` method contains an additional `raw` field requested by @farmaazon for debugging purposes. The `raw` field contains the contents of `package.yaml` as-is. This field should be removed once the integration is complete and the package syntax is stabilized.

To get the available components, one can either:
- Use `editions/listDefinedComponents` method that returns a ready-to-use list of component groups defined by the libraries in the requested edition. 
- Obtain a list of libraries of interest (available methods are `library/listLocal` and `editions/listDefinedLibraries`), and then get the package configs using the `library/getPackage` command.


<!--
- Mention important elements of the design.
- Mention any notable changes to APIs.
-->

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed: Enso GUI was tested when built using BOTH `./run dist` and `./run watch`.

4e6 committed:
feat: getMetadata resolves cached libraries

4e6 committed:
feat: getPackage request

4e6 committed:
feat: listDefinedComponents

4e6 committed:
misc: refactor LibraryRoot

4e6 committed:
refactor: fetchManifestFile

4e6 committed:
test: fix

4e6 committed:
test: jsonrpc

4e6 committed:
test: component groups resolver

4e6 committed:
misc: format

4e6 committed:
misc: document protocol messages

4e6 committed:
feat: return raw package

4e6 committed:
misc: udpate RELEASES

radeusgd commented on docs/language-server/protocol-language-server.md:
```diff
@@ -1391,6 +1400,112 @@ interface NamedEdition {
+
+  /** The list of component groups that this package extends.*/
+  extendedGroups: ExtendedComponentGroup[];
+}
```
If the fields can be ommitted, shouldn't they be marked with the `?` ?

4e6 replied:
Yes, you're right. I thought that `?` indicates nullable fields in TypeScript, but turned out it also indicates an absence of a field.

radeusgd commented on docs/language-server/protocol-language-server.md:
````diff
@@ -1391,6 +1400,112 @@ interface NamedEdition {
+The definition of a component group that extends an existing one.
+
+```typescript
+interface ComponentGroup {
+  /** The reference to the component group module being extended. */
+  module: ModuleReference;
+
+  color?: string;
+
+  icon?: string;
````
This is not directly related to this PR but out of curiosity - why can we specify the icon/color here if all we are doing is extending an existing component group? Won't it be ignored as the ones from the base component group would be used? Or am I misunderstanding something?

4e6 replied:
I thought that is from the design document. But when I checked again, it seems that you should not be able to override the color and icon of an existing group. I'll remove those fields.

radeusgd commented on docs/language-server/protocol-language-server.md:
```diff
@@ -1391,6 +1400,112 @@ interface NamedEdition {
+  libraryName: string;
+
+  /** The module name. */
+  moduleName: string;
```
May be worth clarifying if the `moduleName` contains the `libraryName` as prefix or if it is cut as it is implicit (probably the latter).

radeusgd commented on docs/language-server/protocol-language-server.md:
````diff
@@ -1391,6 +1400,112 @@ interface NamedEdition {
+The definition of a component group that extends an existing one.
+
+```typescript
+interface ComponentGroup {
````
```suggestion
interface ExtendedComponentGroup {
```

radeusgd commented on docs/language-server/protocol-language-server.md:
````diff
@@ -1391,6 +1400,112 @@ interface NamedEdition {
+The component group provided by a library.
+
+```typescript
+interface ComponentGroup {
````
```suggestion
interface LibraryComponentGroup {
```
Shouldn't the name here be consistent with the header?

radeusgd commented on docs/language-server/protocol-language-server.md:
```diff
@@ -4419,6 +4563,47 @@ null;
+from the library repository. A cached package config may also be used, if it is
+available.
+
+All returned fields are optional, as they may be missing.
```
If all are missing, probably `license` should also have a question mark.

radeusgd commented on engine/language-server/src/main/scala/org/enso/languageserver/libraries/ComponentGroupsResolver.scala:
```diff
@@ -0,0 +1,194 @@
+      .collect { case (libraryName, Right(componentGroups)) =>
+        libraryName -> componentGroups
+      }
+    val libraryComponentsMap =
+      ComponentGroupsResolver
+        .toMapKeepFirst(libraryComponents)
```
Why is the `keep first` needed? Can we ever find two configs for the same library name? Maybe in such a situation we should at least log a warning that another config was ignored?

radeusgd commented on engine/language-server/src/main/scala/org/enso/languageserver/libraries/ComponentGroupsResolver.scala:
```diff
@@ -0,0 +1,194 @@
+      libraryComponents.view
+        .flatMap { case (_, componentGroups) =>
+          componentGroups.extendedGroups
+        }
```
Hm not sure how important that is - probably best to checkout in the IDE and then it doesn't matter, but feels like adding type signatures to these 3 definitions would make it easier to understand what is happening here when reading the diffs.

radeusgd commented on engine/language-server/src/main/scala/org/enso/languageserver/libraries/ComponentGroupsResolver.scala:
```diff
@@ -0,0 +1,194 @@
+      val k = f(v)
+      if (m.contains(k)) m
+      else m += k -> v
+    }
```
This just feels so weird that a mutable map is being constructed by a fold.
If it's a fold - maybe it should use an immutable structure?
If using a mutable hashmap is faster, why not just use `each` here or use a `for` loop? Folding suggests that the passed through `m` changes its value - but as a reference to a mutable object, it itself remains constant so there's no need to pass it through.

radeusgd replied:
Also - do we need to expose that this map is mutable? Maybe when it is returned it could be upcast to `collection.Map` (or `MapView`) which hides the mutability?

radeusgd commented on engine/language-server/src/main/scala/org/enso/languageserver/libraries/ComponentGroupsResolver.scala:
```diff
@@ -0,0 +1,194 @@
+        }
+    val newLibraryComponentGroupsMap =
+      ComponentGroupsResolver
+        .groupByKeepFirst(newLibraryComponentGroups) { libraryComponentGroup =>
```
Again - is it ever possible that the `ModuleReference` will be duplicated?
We already deduplicated the libraries earlier, so at this stage a library would need to declare duplicated component groups, right? If this is even allowed, shouldn't a warning be reported somewhere?

It feels like silently ignoring entries like these may lead to hard to detect errors somewhere down the line (but maybe in this particular case I'm misunderstanding something?)

radeusgd commented on engine/language-server/src/main/scala/org/enso/languageserver/libraries/ComponentGroupsResolver.scala:
```diff
@@ -0,0 +1,194 @@
+          .foldLeft(libraryComponentGroup)(applyExtendedComponentGroup)
+        libraryComponentGroups += module -> updatedComponentGroup
+      }
+    }
```
This just feels weird - why can't we just do a `mapValues` like on an immutable map?

radeusgd commented on engine/language-server/src/main/scala/org/enso/languageserver/libraries/ComponentGroupsResolver.scala:
```diff
@@ -0,0 +1,194 @@
+  ): LibraryComponentGroup = {
+    libraryComponentGroup.copy(
+      color   = extendedComponentGroup.color.orElse(libraryComponentGroup.color),
+      icon    = extendedComponentGroup.icon.orElse(libraryComponentGroup.icon),
```
Shouldn't the original color/icon take precedence over the ones defined by the extension? Or is it deliberate that the extension will have the override? (Although that is weird, someone can create a library which will change the color and icon of some base component groups from stdlib? won't that be confusing to the users?)

Also here - the result will depend on the order in which these are resolved - but is the order of `packages: Iterable[Config]` (on which the order of these will rely) at the input specified in any way in `run`? Maybe we should sort the `packages` by name or something - to ensure that the order is static / consistent between runs and platforms (for example if the `packages` is constructed by listing filesystem directories, I think the order in which they would be returned can easily vary between operating systems - potentially leading to component groups looking differently based on the OS).

radeusgd commented on engine/language-server/src/main/scala/org/enso/languageserver/libraries/LibraryApi.scala:
```diff
@@ -163,6 +178,30 @@ object LibraryApi {
+    case class Result(
+      license: String,
+      componentGroups: Option[ComponentGroups],
+      raw: JsonObject
```
I'm wondering if it would be worth to add something like `temporary` to the name of this variable to more directly indicate that it is going to disappear in a while.

radeusgd commented on engine/language-server/src/main/scala/org/enso/languageserver/libraries/LocalLibraryManager.scala:
```diff
@@ -190,6 +192,21 @@ class LocalLibraryManager(
+      config <- loadPackageConfig(configPath)
+    } yield GetPackageResponse(
+      license         = config.license,
+      componentGroups = config.componentGroups.toOption,
```
It will be an option if it failed to parse, right? (If it parsed correctly it will be an empty list instead, right?)

I don't like that we are swallowing errors. I'd suggest at least logging a warning here, specifying the parse error.
But ideally - the error should be forwarded to the IDE as part of the response payload (i.e. for this specific library instead of having a missing field, there could be some other field indicating that this failed to parse; the bulk operation that parses all libraries could include a list of modules which failed to parse, ideally each with the associated error).

I understand that this adds work and is not crucial for an MVP, so maybe it does not need to be done as part of this task - but I would at least suggest to create a task to deal with it at some point. I think it's pretty important to _not_ swallow errors, as this will become a source of confusion for the users at some point: imagine the scenario where a user edits the component group in such a way that it fails to parse - they will stop seeing the respective component groups but will otherwise get no indication of error.

radeusgd commented on engine/language-server/src/main/scala/org/enso/languageserver/libraries/handler/EditionsListDefinedComponentsHandler.scala:
```diff
@@ -0,0 +1,124 @@
+            .map { case (name, version) =>
+              readLocalPackage(name, version)
+            }
+            .collect { case Some(config) =>
+              config
+            }
```
Again, shouldn't libraries for which the config could not be loaded be indicated somehow? Just ignoring them feels a bit bad. If not in this iteration, at least a TODO could be left for the future - otherwise I'd think it will be unlikely this gets revisited.

radeusgd commented on engine/language-server/src/main/scala/org/enso/languageserver/libraries/handler/LibraryGetPackageHandler.scala:
```diff
@@ -0,0 +1,175 @@
+  * @param localLibraryManager reference to the local library manager actor
+  * @param publishedLibraryCache the cache of published libraries
+  */
+class LibraryGetPackageHandler(
```
The implementation of this class looks so similar to `LibraryGetMetadataHandler` that I'm wondering if maybe it would be worth to separate out the shared logic to avoid repeating it?

radeusgd commented on engine/language-server/src/test/scala/org/enso/languageserver/libraries/ComponentGroupsResolverSpec.scala:
```diff
@@ -0,0 +1,258 @@
+      )
+    }
+
+    "skip component groups extending nothing" in {
```
Again, shouldn't this be reported somewhere as a warning?

If I make a typo in specifying what existing group I want to extend, my extensions will just not show up - not very helpful when trying to figure out what was the issue. Not saying this is necessary from the beginning, but at least an indication that it should be added at some point would IMO be good.

radeusgd commented on engine/language-server/src/test/scala/org/enso/languageserver/websocket/json/LibrariesTest.scala:
```diff
@@ -609,6 +715,55 @@ class LibrariesTest extends BaseServerTest {
+          """)
+
+      val currentEditionName = buildinfo.Info.currentEdition
+      client.send(json"""
+          { "jsonrpc": "2.0",
+            "method": "editions/listDefinedComponents",
+            "id": 0,
+            "params": {
+              "edition": {
+                "type": "NamedEdition",
+                "editionName": $currentEditionName
+              }
+            }
+          }
+          """)
+
+      client.expectJson(json"""
+          { "jsonrpc": "2.0",
+            "id": 0,
+            "result": {
+              "availableComponents" : [ ]
+            }
+          }
```
Maybe you could somehow fake the current edition to have any component groups? Feels like the test is a bit incomplete if only the path leading to 'no results' is tested.

radeusgd commented on lib/scala/library-manager/src/main/scala/org/enso/librarymanager/published/CachedLibraryProvider.scala:
```diff
@@ -1,27 +1,24 @@
 /** A [[PublishedLibraryProvider]] that just provides libraries which are
   * already available in the cache.
   */
 class CachedLibraryProvider(caches: List[ReadOnlyLibraryCache])
-    extends PublishedLibraryProvider
-    with PublishedLibraryCache {
+    extends PublishedLibraryCache {
```
This comment is not valid anymore - this class no longer implements the `PublishedLibraryProvider` interface, like suggested in the doc comment.

radeusgd reviewed (changes requested):
Overall - looks really great; it took me quite a while to review this as this is adds quite a lot of functionality.

1. Some comments on documentation that I think should be addressed before merging (inconsistent names etc.)

Then some lower priority stuff but would be cool if addressed:
2. As written in comments above - in multiple places we swallow errors regarding parsing and handling of component groups. This seems bad to me - the users will be confused why their defined groups do not show up without getting any cues. While I understand this may not be the most pressing feature at this stage - I think it would be good to at least mark these relevant places with a TODO and adding a task to address this - so that we can keep track of the places that will need to be revisited to have proper error handling.
3. Some weird style where a mutable data structure is handling using operations associated with immutable structures (folding where the accumulator is a mutable map) - while not incorrect, it just looks strange (that is a bit subjective though). I'd suggest maybe changing this to be more uniform - either by using immutable structures or by writing more imperative code using mutable structures - the mixup just looks a bit weird.
4. If possible, would be cool to integration test the `editions/listDefinedComponents` endpoint - as (unless I missed something which is not out of question) seems like is only tested with the scenario where no components are returned - which seems a bit lacking.

4e6 committed:
update language-server api

4e6 committed:
feat: validate component groups

4e6 committed:
doc: misc

4e6 commented:
@radeusgd thanks for the review. I managed to address most of your comments.
1. API is adjusted. I removed redundant fields, fixed the names
2. Implemented config verification and error logging. In the future can extend it and send the verification errors to the client
3. I changed the data structures returned from grouping combinators to immutable ones. Although I believe that folding with the mutable reference is fine, because it does not affect the referential transparency (you will still get the same result as with immutable one)
4. I was not able to mock the edition in order to implement a proper test for the `listDefinedEditions`. The result of this test will be filled with component groups when we add them to the standard library

4e6 committed:
feat: log component group parsing error

radeusgd reviewed (approved):
Awesome!

> 2. Implemented config verification and error logging. In the future can extend it and send the verification errors to the client

Looks perfect. I think we should add a way to send the errors to the client, but totally understandable to make this a separate task (and probably not the top-priority). Do you think it's worth creating a task in Pivotal to at least keep it somewhere in the Icebox?

> 3. I changed the data structures returned from grouping combinators to immutable ones. Although I believe that folding with the mutable reference is fine, because it does not affect the referential transparency (you will still get the same result as with immutable one)

I totally agree that folding a mutable reference is fine - in the sense that it is correct. It is just an uncommon pattern which IMO may be confusing when reading it (mixing imperative and functional constructs like this). But that is just a subjective code-style opinion, so I'm not pushing it.

> 4. I was not able to mock the edition in order to implement a proper test for the `listDefinedEditions`. The result of this test will be filled with component groups when we add them to the standard library

Yeah, I expected this may be problematic. Would be good to have it, but if the component groups are going to be added to the standard library relatively soon, I guess we are good (not much value in implementing a mock which will not be necessary soon).
Would be cool to temporarily add a component group to stdlib and at least test it manually/locally - just as a sanity check. Unless you already did that :)

4e6 committed:
Merge branch 'develop' into wip/db/component-api
